model/task_queue_service_dependency.go

```diff
@@ -30,139 +30,230 @@ type basicCachedDAGDispatcherImpl struct {
-	node := t.graph.NewNode()
-	t.graph.AddNode(node)
-	t.nodeItemMap[node.ID()] = item
-	t.itemNodeMap[item.Id] = node
+func (d *basicCachedDAGDispatcherImpl) addItem(item *TaskQueueItem) {
+	node := d.graph.NewNode()
+	d.graph.AddNode(node)
+	d.nodeItemMap[node.ID()] = item
+	d.itemNodeMap[item.Id] = node
 }
 
-func (t *basicCachedDAGDispatcherImpl) getItemByNodeID(id int64) *TaskQueueItem {
-	if item, ok := t.nodeItemMap[id]; ok {
+func (d *basicCachedDAGDispatcherImpl) getItemByNodeID(id int64) *TaskQueueItem {
+	if item, ok := d.nodeItemMap[id]; ok {
 		return item
 	}
 
 	return nil
 }
 
-func (t *basicCachedDAGDispatcherImpl) getNodeByItemID(id string) graph.Node {
-	if node, ok := t.itemNodeMap[id]; ok {
+func (d *basicCachedDAGDispatcherImpl) getNodeByItemID(id string) graph.Node {
+	if node, ok := d.itemNodeMap[id]; ok {
 		return node
 	}
 
 	return nil
 }
 
-// Each node is a task and each edge definition represents a dependency; an edge (A, B) means that B depends on A.
-func (t *basicCachedDAGDispatcherImpl) addEdge(from string, to string) error {
-	fromNode := t.getNodeByItemID(from)
-	toNode := t.getNodeByItemID(to)
+// Each node is a task and each edge definition represents a dependency: an edge (A, B) means that B depends on A.
+// There is a dependency <from> A <to> B.
+func (d *basicCachedDAGDispatcherImpl) addEdge(fromID string, toID string) error {
+	fromNode := d.getNodeByItemID(fromID)
+	toNode := d.getNodeByItemID(toID)
 
 	if fromNode == nil {
-		return errors.Errorf("a Node for taskQueueItem '%s' is not present in the DAG for distro '%s'", from, t.distroID)
+		// A Node for the "depends_on" <from> task to be satisfied is not present in the DAG - why?
+		//
+		// Maybe...
+		//
+		//	(1)	Task A completes (evergreen.TaskSucceeded, evergreen.TaskFailed) and some tasks that "depend_on" it are still enqueued behind it.
+		//			- Some time passes (determined by a ttl value) and a new task_queue is created.
+		//			- However, other taskQueueItems in the latest task_queue may "depend_on" it and it is not in the latest queue_task.
+		//			- [TBC] But, shouldn't something be settings the dependsOn._id.Unattainable = true for all tasks that "depend_on" a failed task? Is there a race condition?
+		// 	(2)	[TBC] The "dependent" <to> task is explicitly set to override its dependencies via task.OverrideDependencies.
+		//	(3)	The "depends_on" <from> task is itself a "dependent" task with its own set of "depends_on" tasks to satisfy.
+		// 		 	- [TBC] There is a recursion depth of 0 when generating the set of tasks ("dependent" plus "depends_on") that will comprise the task_queue to be dispatched.
+		//
+		// Regardless, for now we will assume the task_finder has done the right thing and continue debugging.
+
+		// Get the "dependent" <to> task from the database.
+		toTask, err := task.FindOneId(toID)
+		if err != nil {
+			grip.Error(message.WrapError(err, message.Fields{
```

Comment:
maybe put all of this logging at level.Warning, and put a constant in for the dispatcher value (to avoid copying) and then set up a splunk query/alert for yourself (+me?) to see if we hit any of these problems.

Reply:
Logging set to level.Warning.